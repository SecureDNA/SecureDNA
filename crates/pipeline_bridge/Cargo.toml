[package]
name = "pipeline_bridge"
version = { workspace = true }
publish = { workspace = true }
license = { workspace = true }
edition = "2021"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
serde = { version = "1.0", features = ["derive"] }
smartstring = { version = "1.0.1", features = ["serde"] }

[dev-dependencies]
version_check = { version = ">= 0.9.4" } # fix smartstring build: https://github.com/bodil/smartstring/issues/30#issuecomment-1080194045
