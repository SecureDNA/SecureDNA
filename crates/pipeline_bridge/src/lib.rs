// Copyright 2021-2024 SecureDNA Stiftung (SecureDNA Foundation) <licensing@securedna.org>
// SPDX-License-Identifier: MIT OR Apache-2.0

//! Crate for interfacing with the output of the pipeline.
//!
//! The pipeline produces artifacts which genhdb processes into an hdb. This create provides types
//! so that those artifacts can be easily deserialized.
//!
//! Note that these types are parallel types to those used by the pipeline aggregator to serialize
//! the artifacts. So make sure to keep the types in sync (they are similar but not identical)

use std::path::PathBuf;

use serde::{Deserialize, Serialize};
use smartstring::{LazyCompact, SmartString};

pub const PROTEIN_LEN: usize = 20;
pub const DNA_NORMAL_LEN: usize = 42;
pub const DNA_RUNT_LEN: usize = 30;

#[derive(
    Debug,
    Copy,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    std::hash::Hash,
    serde::Serialize,
    serde::Deserialize,
)]
// tsgen
pub enum OrganismType {
    Virus,
    Toxin,
    Bacterium,
    Fungus,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct HazardProperties {
    pub path_name: String,
    pub common_name: String,
    pub accessions: Vec<String>,
    pub tags: Vec<Tag>,
    // Panics if unable to deserialize
    pub organism_type: OrganismType,
    // Defaults to false (shingled)
    #[serde(default)]
    pub tiled: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
pub enum Provenance {
    #[serde(rename = "wild_type")]
    WildType,
    #[serde(rename = "initial_1_variant")]
    SingleReplacement,
    #[serde(rename = "initial_2_variant")]
    DoubleReplacement,
    #[serde(rename = "mh_sample")]
    MHSample,
}

/// Genhdb doesn't need any of the fields of this, just the existence
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct ReverseScreeningHit {}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct VariantEntry {
    pub variant: SmartString<LazyCompact>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provenance: Option<Provenance>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub log_likelihood: Option<f32>,
    /// If Some, this window matched in reverse screening with the given MatchedWindow
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reverse_screened: Option<ReverseScreeningHit>,
    // If variant is a common seq
    pub is_common: bool,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct AggregatedHazard {
    /// Metadata for this hazard
    pub hazard_meta: HazardProperties,
    /// Path (relative to artifacts dir) to fraglist file (one VariantEntry per line)
    pub dna_variant_42mers_path: PathBuf,
    /// Path (relative to artifacts dir) to fraglist file (one VariantEntry per line)
    pub dna_variant_30mers_path: PathBuf,
    /// Path (relative to artifacts dir) to fraglist file (one VariantEntry per line)
    pub protein_variants_path: PathBuf,
}

#[derive(
    Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, std::hash::Hash, Deserialize, Serialize,
)]
/// Tags for hazard metadata.
/// A tag may represent
/// - transmission pathway
/// - risk categorization
/// - inclusion in a controlled hazard list published by an organization
/// - internal metadata, such as whether to generate runts or how to treat organism matches
///   (these tags should still not include sensitive information, see lib.dhall)
///
///   Note: Please keep in alphabetical order, this matters for tests
// tsgen
pub enum Tag {
    /// Arthropod-to-human transmissible
    ArthropodToHuman,
    /// Australia group human animal pathogens
    /// <https://www.dfat.gov.au/publications/minisite/theaustraliagroupnet/site/en/human_animal_pathogens.html>
    AustraliaGroupHumanAnimalPathogen,
    /// Australia group plant pathogens
    /// <https://www.dfat.gov.au/publications/minisite/theaustraliagroupnet/site/en/plants.html>
    AustraliaGroupPlantPathogen,
    /// Whether the tagged sequence is a common sequence
    Common,
    /// European Union
    EuropeanUnion,
    /// Human-to-human transmissible
    HumanToHuman,
    /// Potential pandemic pathogen
    PotentialPandemicPathogen,
    /// People's Republic of China export control list (2002), parts 1 and 2
    PRCExportControlPart1,
    /// People's Republic of China export control list (2002), parts 1 and 2
    PRCExportControlPart2,
    // Generated tag; not sure if it should be in this enum (which is basically what is sent from the
    // pipelin), or if there should be a superset of this `Tag` for generated tags.
    RegulatedButPass,
    /// Internal tag: treat DNA matches from this organism specially in the HDB
    /// Intended for use with "regulatory compliance only" genomes like E. coli
    SdnaLowRiskDNA,
    /// Internal tag: normally we exempt virus wild types from reverse screening
    /// (by treating all RS matches against virus wild types as self matches),
    /// but for certain large viruses we don't want to do that. This tag turns off that
    /// behavior so the virus wild type will be reverse screened like virus variants.
    SdnaForceVirusReverseScreening,
    /// USA Select agent: US Department of Health and Human Services
    SelectAgentHhs,
    /// USA Select agent: US Department of Agriculture
    SelectAgentUsda,
    /// USA Select agent: USDA Animal and Plant Health Inspection Service
    SelectAgentAphis,
}

impl Tag {
    /// Function that returns whether a tag starts with any of the [`INTERNAL_TAG_PREFIXES`]()
    /// Internal tags won't be returned in the API response.
    /// We still shouldn't put anything extremely sensitive here, in case of HLT leakage or accidental tag returnâ€”this is more because customers don't care about these tags, not because they're dangerous to reveal. For example, we will only turn off runt generation for organisms that don't need it for security, so if the NoRuntGeneration tag leaked, it wouldn't matter.
    ///
    /// We don't check prefixes, instead we match exactly on the tag. This is because we
    /// maintain an exhaustive list of internal tags anyways.
    pub fn is_internal(&self) -> bool {
        match self {
            Tag::SdnaLowRiskDNA | Tag::SdnaForceVirusReverseScreening => true,
            Tag::ArthropodToHuman
            | Tag::AustraliaGroupHumanAnimalPathogen
            | Tag::AustraliaGroupPlantPathogen
            | Tag::Common
            | Tag::EuropeanUnion
            | Tag::HumanToHuman
            | Tag::PotentialPandemicPathogen
            | Tag::PRCExportControlPart1
            | Tag::PRCExportControlPart2
            | Tag::RegulatedButPass
            | Tag::SelectAgentHhs
            | Tag::SelectAgentUsda
            | Tag::SelectAgentAphis => false,
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct BuildTrace {
    pub build_timestamp: String,
    pub pipeline_git_sha: String,
    pub pipeline_git_timestamp: String,
    pub hdb_git_sha: String,
    pub hlt_tags_modified_git_sha: Option<String>,
    pub previous_build_info: Option<Box<BuildTrace>>,
}
