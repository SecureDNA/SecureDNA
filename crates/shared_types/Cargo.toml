[package]
name = "shared_types"
version = { workspace = true }
publish = { workspace = true }
license = { workspace = true }
edition = "2021"

[dependencies]
base64 = "0.22.0"
http = { version = "1.0.0", optional = true }
once_cell = "1.19.0"
prometheus = "0.13.3"
regex = "1.10.3"
serde = { workspace = true, features = ["derive"] }
serde_plain = "1.0.1"
thiserror = "1.0.47"
uuid = {version = "1.10.0", features = ["v4"]}

certificates = { path = "../certificates" }
doprf = {path = "../doprf"}
pipeline_bridge = { path = "../pipeline_bridge" }

[dev-dependencies]
serde_json = "1.0"
quickcheck = "1.0"
