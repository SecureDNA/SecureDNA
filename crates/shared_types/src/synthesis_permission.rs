// Copyright 2021-2024 SecureDNA Stiftung (SecureDNA Foundation) <licensing@securedna.org>
// SPDX-License-Identifier: MIT OR Apache-2.0

use serde::{Deserialize, Serialize};

use pipeline_bridge::Tag;

/// Region jurisdictions for handling requests. Controls e.g. what rules to use for setting
/// the `synthesis_permission` bit to `denied`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Region {
    /// United States
    Us,
    /// European Union
    Eu,
    /// People's Republic of China
    Prc,
    /// Check all regions. This is the default. It means:
    ///
    /// - Synthesis is granted only if the organism is safe in all regions.
    /// - Synthesis is denied if the organism is controlled in *any* region.
    All,
}

impl Default for Region {
    fn default() -> Self {
        Self::All
    }
}

serde_plain::derive_fromstr_from_deserialize!(Region);
serde_plain::derive_display_from_serialize!(Region);

#[derive(Debug, Clone, Copy, PartialEq, Eq, std::hash::Hash, Serialize, Deserialize)]
pub enum SynthesisPermission {
    #[serde(rename = "granted")]
    Granted,
    #[serde(rename = "denied")]
    Denied,
}

impl SynthesisPermission {
    /// Merge multiple `SynthesisPermission`s into one, returning
    /// `SynthesisPermission::Denied` if any are denials.
    ///
    /// Returns `SynthesisPermission::Granted` if the iterator is empty.
    pub fn merge<I, SP>(perms: I) -> Self
    where
        I: IntoIterator<Item = SP>,
        SP: AsRef<SynthesisPermission>,
    {
        let any_denied = perms.into_iter().any(|p| *p.as_ref() == Self::Denied);

        if any_denied {
            Self::Denied
        } else {
            Self::Granted
        }
    }
}

impl AsRef<SynthesisPermission> for SynthesisPermission {
    fn as_ref(&self) -> &SynthesisPermission {
        self
    }
}

impl From<SynthesisPermission> for &'static str {
    fn from(value: SynthesisPermission) -> Self {
        match value {
            SynthesisPermission::Granted => "granted",
            SynthesisPermission::Denied => "denied",
        }
    }
}

pub fn permission_for_region(
    tags: impl IntoIterator<Item = Tag>,
    region: Region,
) -> SynthesisPermission {
    let permission_by_tag = tags
        .into_iter()
        .map(|tag| tag_by_region_flag_table(tag, region));
    SynthesisPermission::merge(permission_by_tag)
}

pub fn tag_by_region_flag_table(tag: Tag, region: Region) -> SynthesisPermission {
    match tag {
        Tag::PotentialPandemicPathogen => SynthesisPermission::Denied,
        Tag::SelectAgentHhs => match region {
            Region::Us | Region::All => SynthesisPermission::Denied,
            Region::Prc | Region::Eu => SynthesisPermission::Granted,
        },
        Tag::SelectAgentUsda => match region {
            Region::Us | Region::All => SynthesisPermission::Denied,
            Region::Prc | Region::Eu => SynthesisPermission::Granted,
        },
        Tag::SelectAgentAphis => match region {
            Region::Us | Region::All => SynthesisPermission::Denied,
            Region::Prc | Region::Eu => SynthesisPermission::Granted,
        },
        Tag::AustraliaGroupHumanAnimalPathogen => match region {
            Region::Us | Region::Eu | Region::All => SynthesisPermission::Denied,
            Region::Prc => SynthesisPermission::Granted,
        },
        Tag::AustraliaGroupPlantPathogen => match region {
            Region::Us | Region::Eu | Region::All => SynthesisPermission::Denied,
            Region::Prc => SynthesisPermission::Granted,
        },
        Tag::PRCExportControlPart1 => match region {
            Region::Prc | Region::All => SynthesisPermission::Denied,
            Region::Us | Region::Eu => SynthesisPermission::Granted,
        },
        Tag::PRCExportControlPart2 => match region {
            Region::Prc | Region::All => SynthesisPermission::Denied,
            Region::Us | Region::Eu => SynthesisPermission::Granted,
        },
        Tag::EuropeanUnion => match region {
            Region::Eu | Region::All => SynthesisPermission::Denied,
            Region::Us | Region::Prc => SynthesisPermission::Granted,
        },
        Tag::Common
        | Tag::HumanToHuman
        | Tag::RegulatedButPass
        | Tag::SdnaLowRiskDNA
        | Tag::SdnaForceVirusReverseScreening
        | Tag::ArthropodToHuman => SynthesisPermission::Granted,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // Test that the big match statement is consistent with the human-readable table
    const TABLE: &str = r#"
    ----------------------------------------------------------------------
    - Tag                             | Us | Eu | Prc | All (no region provided)
    ----------------------------------------------------------------------
    SelectAgentHhs                    | X  | -  | -   | X
    SelectAgentUsda                   | X  | -  | -   | X
    SelectAgentAphis                  | X  | -  | -   | X
    AustraliaGroupHumanAnimalPathogen | X  | X  | -   | X
    AustraliaGroupPlantPathogen       | X  | X  | -   | X
    PRCExportControlPart1             | -  | -  | X   | X
    PRCExportControlPart2             | -  | -  | X   | X
    EuropeanUnion                     | -  | X  | -   | X
    ArthropodToHuman                  | -  | -  | -   | -
    HumanToHuman                      | -  | -  | -   | -
    PotentialPandemicPathogen         | X  | X  | X   | X
    RegulatedButPass                  | -  | -  | -   | -
    -------------------------------------------------------------
    "#;

    #[test]
    fn test_tag_table() {
        for line in TABLE.split('\n') {
            let line = line.trim();
            if line.is_empty() || line.starts_with('-') {
                continue;
            }
            let parts = line.split('|').map(|s| s.trim()).collect::<Vec<_>>();
            let [tag_name, match_us, match_eu, match_prc, match_no_region] = parts[..] else {
                panic!("bad line format: {parts:?}");
            };
            let tag: Tag = serde_json::from_str(&format!("\"{tag_name}\"")).unwrap();

            fn to_sp(s: &str) -> SynthesisPermission {
                match s {
                    "X" => SynthesisPermission::Denied,
                    "-" => SynthesisPermission::Granted,
                    other => panic!("expected X or -, got {other:?}"),
                }
            }

            assert_eq!(permission_for_region([tag], Region::Us), to_sp(match_us),);
            assert_eq!(permission_for_region([tag], Region::Eu), to_sp(match_eu),);
            assert_eq!(permission_for_region([tag], Region::Prc), to_sp(match_prc),);
            assert_eq!(
                permission_for_region([tag], Region::All),
                to_sp(match_no_region)
            );
        }
    }

    #[test]
    fn merge_empty_granted() {
        let empty: &[SynthesisPermission] = &[];
        assert_eq!(
            SynthesisPermission::merge(empty),
            SynthesisPermission::Granted
        );
    }

    #[test]
    fn merge_granted() {
        assert_eq!(
            SynthesisPermission::merge([
                SynthesisPermission::Granted,
                SynthesisPermission::Granted
            ]),
            SynthesisPermission::Granted,
        )
    }

    #[test]
    fn merge_denied() {
        assert_eq!(
            SynthesisPermission::merge([SynthesisPermission::Granted, SynthesisPermission::Denied]),
            SynthesisPermission::Denied,
        )
    }
}
