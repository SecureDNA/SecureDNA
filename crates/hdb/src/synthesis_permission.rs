// Copyright 2021-2024 SecureDNA Stiftung (SecureDNA Foundation) <licensing@securedna.org>
// SPDX-License-Identifier: MIT OR Apache-2.0

use crate::{Exemptions, HdbOrganism};
use pipeline_bridge::Tag;
use shared_types::synthesis_permission::{permission_for_region, Region, SynthesisPermission};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct PermissionResult {
    /// Whether permission was actually granted or denied.
    pub permission: SynthesisPermission,
    /// If true, this means the permission was granted, but would have been
    /// denied were it not for the provided EL.
    pub exempt: bool,
}

impl PermissionResult {
    pub fn granted() -> Self {
        Self {
            permission: SynthesisPermission::Granted,
            exempt: false,
        }
    }

    pub fn exempt() -> Self {
        Self {
            permission: SynthesisPermission::Granted,
            exempt: true,
        }
    }

    pub fn denied() -> Self {
        Self {
            permission: SynthesisPermission::Denied,
            exempt: false,
        }
    }
}

/// Calculate `synthesis_permission` based on an incomplete HDB response
pub fn get_permission<'o>(
    organisms: impl IntoIterator<Item = &'o HdbOrganism>,
    reverse_screened: bool,
    region: Region,
    exemptions: &Exemptions,
) -> PermissionResult {
    use SynthesisPermission::{Denied, Granted};

    if reverse_screened {
        return PermissionResult::granted();
    }

    let mut all_tags: Vec<Tag> = vec![];
    let mut actual_tags: Vec<Tag> = vec![];

    for organism in organisms.into_iter() {
        let exempt = exemptions.is_organism_exempt(organism);
        all_tags.extend(&organism.tags);
        if !exempt {
            actual_tags.extend(&organism.tags);
        }
    }

    let actual_permission = permission_for_region(actual_tags, region);
    let without_el = permission_for_region(all_tags, region);
    match (actual_permission, without_el) {
        (Granted, Granted) => PermissionResult::granted(),
        (Granted, Denied) => PermissionResult::exempt(),
        (Denied, _) => PermissionResult::denied(),
    }
}

#[cfg(test)]
mod tests {
    use certificates::{GenbankId, SequenceIdentifier};
    use pipeline_bridge::{OrganismType, Tag};

    use super::*;

    #[test]
    fn empty_hits_granted() {
        assert_eq!(
            get_permission(&[], false, Region::All, &Default::default()),
            PermissionResult::granted()
        );
        assert_eq!(
            get_permission(&[], true, Region::All, &Default::default()),
            PermissionResult::granted()
        );
    }

    fn make_organism(tags: &[Tag]) -> HdbOrganism {
        HdbOrganism {
            name: "Testus Unitae".into(),
            organism_type: OrganismType::Virus,
            ans: vec!["AN_1234".into()],
            tags: tags.into(),
        }
    }

    fn make_exemptions() -> Exemptions {
        crate::exemption::make_test_exemptions(vec![certificates::Organism {
            name: "Testus Unitae".into(),
            sequences: vec![SequenceIdentifier::Id(
                GenbankId::try_from("AN_1234".to_owned()).unwrap(),
            )],
        }])
    }

    #[test]
    fn some_hits_denied() {
        assert_eq!(
            get_permission(
                &[make_organism(&[Tag::SelectAgentAphis])],
                false,
                Region::All,
                &Default::default()
            ),
            PermissionResult::denied(),
        )
    }

    #[test]
    fn rs_granted() {
        assert_eq!(
            get_permission(
                &[make_organism(&[Tag::SelectAgentAphis])],
                true,
                Region::All,
                &Default::default()
            ),
            PermissionResult::granted(),
        )
    }

    #[test]
    fn benign_tag_granted() {
        assert_eq!(
            get_permission(
                &[make_organism(&[Tag::ArthropodToHuman])],
                false,
                Region::All,
                &Default::default()
            ),
            PermissionResult::granted(),
        )
    }

    #[test]
    fn diff_region_tag_granted() {
        assert_eq!(
            get_permission(
                &[make_organism(&[Tag::PRCExportControlPart1])],
                false,
                Region::Us,
                &Default::default()
            ),
            PermissionResult::granted(),
        )
    }

    #[test]
    fn same_region_tag_denied() {
        assert_eq!(
            get_permission(
                &[make_organism(&[Tag::PRCExportControlPart1])],
                false,
                Region::Prc,
                &Default::default()
            ),
            PermissionResult::denied(),
        )
    }

    #[test]
    fn exemption_granted() {
        assert_eq!(
            get_permission(
                &[make_organism(&[Tag::SelectAgentAphis])],
                false,
                Region::All,
                &make_exemptions()
            ),
            PermissionResult::exempt(),
        )
    }

    #[test]
    fn region_exemption_granted() {
        assert_eq!(
            get_permission(
                &[make_organism(&[Tag::PRCExportControlPart1])],
                false,
                Region::Prc,
                &make_exemptions()
            ),
            PermissionResult::exempt(),
        )
    }
}
