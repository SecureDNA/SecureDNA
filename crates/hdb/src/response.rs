// Copyright 2021-2024 SecureDNA Stiftung (SecureDNA Foundation) <licensing@securedna.org>
// SPDX-License-Identifier: MIT OR Apache-2.0

use serde::{Deserialize, Serialize};

use pipeline_bridge::{OrganismType, Tag};
use shared_types::synthesis_permission::{Region, SynthesisPermission};

use crate::{
    hlt::HLTLookupError, synthesis_permission::PermissionResult, tags, Exemptions, HLTId,
    HazardLookupTable, Metadata, Provenance,
};

/// Note: if modifying this structure, make sure to also modify Hash impl!
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct HdbResponse {
    pub synthesis_permission: SynthesisPermission,
    pub most_likely_organism: HdbOrganism,
    pub organisms: Vec<HdbOrganism>,
    pub an_likelihood: f32,
    pub provenance: Provenance,
    pub reverse_screened: bool,
    pub window_gap: usize,
    pub exempt: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub struct HdbOrganism {
    pub name: String,
    pub organism_type: OrganismType,
    pub ans: Vec<String>,
    pub tags: Vec<Tag>,
}

impl HdbResponse {
    /// For comparisons w/out an_likelihood, which allows us to sum an_likelihood during
    /// consolidation
    pub fn eq_without_an_likelihood(&self, other: &Self) -> bool {
        self.synthesis_permission == other.synthesis_permission
            && self.most_likely_organism == other.most_likely_organism
            && self.organisms == other.organisms
            && self.provenance == other.provenance
            && self.reverse_screened == other.reverse_screened
            && self.window_gap == other.window_gap
    }
}

#[derive(Debug, Clone, thiserror::Error)]
pub enum HdbResponseError {
    #[error("metadata hlt index {0} missing in provided HLT")]
    MissingMetaHLTIndex(u32),
    #[error("metadata an subindex {1} missing in provided HLT entry {0}")]
    MissingMetaANSubindex(u32, u8),
}

impl HdbOrganism {
    pub fn from_ids(
        ids: &[HLTId],
        hlt_index: u32,
        an_subindex: u8,
    ) -> Result<Self, HLTLookupError> {
        let mut name: Option<String> = None;
        let mut organism_type: Option<OrganismType> = None;
        let mut ans = vec![];
        let mut tags = vec![];

        for id in ids {
            match id {
                HLTId::OrganismName(s) => name = Some(s.clone()),
                HLTId::Accession(an) => ans.push(an.clone()),
                HLTId::Tag(tag) => tags.push(*tag),
                HLTId::Tiled => {}
                HLTId::OrganismType(t) => organism_type = Some(*t),
            }
        }

        let name = name.ok_or(HLTLookupError::MissingOrganismName(hlt_index, an_subindex))?;
        let organism_type =
            organism_type.ok_or(HLTLookupError::MissingOrganismType(hlt_index, an_subindex))?;
        ans.sort();
        ans.dedup();
        tags.sort();
        tags.dedup();
        Ok(HdbOrganism {
            name,
            organism_type,
            ans,
            tags,
        })
    }
}

impl HdbResponse {
    pub fn with_hlt(
        metadata: Metadata,
        region: Region,
        exemptions: &Exemptions,
        hlt: &HazardLookupTable,
    ) -> Result<Self, HLTLookupError> {
        let (hlt_entry, likely_hltids) =
            hlt.get_with_subindex(&metadata.hlt_index, &metadata.an_subindex)?;

        let mut organisms: Vec<HdbOrganism> = hlt_entry
            .iter()
            .map(|ids| HdbOrganism::from_ids(ids.1, metadata.hlt_index, metadata.an_subindex))
            .collect::<Result<_, _>>()?;
        let mut most_likely_organism =
            HdbOrganism::from_ids(likely_hltids, metadata.hlt_index, metadata.an_subindex)?;

        tags::remove_or_transform_internal_tags(
            metadata.provenance,
            &mut most_likely_organism.tags,
            organisms.iter_mut().map(|organism| &mut organism.tags),
        );
        if metadata.is_common {
            most_likely_organism.tags.push(Tag::Common);
            for organism in organisms.iter_mut() {
                organism.tags.push(Tag::Common);
            }
        }

        let an_likelihood: f32 = metadata.an_likelihood.into();
        assert!(!an_likelihood.is_nan());

        // It's unlikely that we'll cross major taxonomic boundaries, so we'd expect that if one
        // hazard is tiled, then all hazards should be tiled (same for hazard_type)
        let tiled = likely_hltids.iter().any(HLTId::tiled);
        let window_gap = metadata.provenance.window_gap(tiled);

        let PermissionResult { permission, exempt } = crate::synthesis_permission::get_permission(
            &organisms,
            metadata.reverse_screened,
            region,
            exemptions,
        );

        Ok(Self {
            synthesis_permission: permission,
            most_likely_organism,
            organisms,
            an_likelihood,
            provenance: metadata.provenance,
            reverse_screened: metadata.reverse_screened,
            window_gap,
            exempt,
        })
    }
}

#[cfg(test)]
mod tests {
    use certificates::{GenbankId, Organism, SequenceIdentifier};

    use crate::exemption::make_exemptions;

    use super::*;

    static COMPLICATED_HLT: &str = r#"
    {
        "entries": {
            "0": { "id_groups": [
                [
                    {"OrganismName": "Nastyitis"},
                    {"OrganismType": "Virus"},
                    {"Accession": "NC_00000.0"},
                    {"Accession": "NC_00000.1"},
                    {"Tag": "HumanToHuman"},
                    {"Tag": "SdnaLowRiskDNA"}
                ],
                [
                    {"OrganismName": "Nastyitis variant"},
                    {"OrganismType": "Virus"},
                    {"Accession": "NC_00001"},
                    {"Tag": "HumanToHuman"},
                    {"Tag": "ArthropodToHuman"},
                    {"Tag": "SdnaLowRiskDNA"}
                ]
            ] },
            "198": { "id_groups": [
                [
                    {"OrganismName": "Nastytoxin"},
                    {"OrganismType": "Toxin"},
                    {"Accession": "NC_00002"},
                    {"Tag": "SelectAgentHhs"},
                    {"Tag": "SdnaLowRiskDNA"}
                ]
            ] }
        }
    }
    "#;

    #[test]
    pub fn generates_correct_response_single_subentry() {
        let hlt: HazardLookupTable = serde_json::from_str(COMPLICATED_HLT).unwrap();
        let metadata = Metadata {
            hlt_index: 198,
            an_subindex: 0,
            an_likelihood: half::f16::from_f32(0.5),
            provenance: Provenance::AAWildType,
            reverse_screened: false,
            is_common: false,
        };

        insta::assert_yaml_snapshot!(HdbResponse::with_hlt(
            metadata,
            Region::Us,
            &Default::default(),
            &hlt
        )
        .unwrap())
    }

    #[test]
    pub fn generates_correct_response_multiple_subentry() {
        let hlt: HazardLookupTable = serde_json::from_str(COMPLICATED_HLT).unwrap();
        let metadata = Metadata {
            hlt_index: 0,
            an_subindex: 0,
            an_likelihood: half::f16::from_f32(0.5),
            provenance: Provenance::AAWildType,
            reverse_screened: false,
            is_common: false,
        };

        insta::assert_yaml_snapshot!(HdbResponse::with_hlt(
            metadata,
            Region::Us,
            &Default::default(),
            &hlt
        )
        .unwrap())
    }

    #[test]
    pub fn generates_correct_response_multiple_subentry_organism_not_likely() {
        let hlt: HazardLookupTable = serde_json::from_str(COMPLICATED_HLT).unwrap();
        let metadata = Metadata {
            hlt_index: 0,
            an_subindex: 1,
            an_likelihood: half::f16::from_f32(0.5),
            provenance: Provenance::AAWildType,
            reverse_screened: false,
            is_common: false,
        };
        insta::assert_yaml_snapshot!(HdbResponse::with_hlt(
            metadata,
            Region::Us,
            &Default::default(),
            &hlt
        )
        .unwrap())
    }

    #[test]
    pub fn generates_correct_response_exemption() {
        let hlt: HazardLookupTable = serde_json::from_str(COMPLICATED_HLT).unwrap();
        let metadata = Metadata {
            hlt_index: 198,
            an_subindex: 0,
            an_likelihood: half::f16::from_f32(0.5),
            provenance: Provenance::AAWildType,
            reverse_screened: false,
            is_common: false,
        };

        let exemption = Organism {
            name: "Nastytoxin".into(),
            sequences: vec![SequenceIdentifier::Id(
                GenbankId::try_from("NC_00002".to_owned()).unwrap(),
            )],
        };

        insta::assert_yaml_snapshot!(HdbResponse::with_hlt(
            metadata,
            Region::Us,
            &make_exemptions(vec![exemption]),
            &hlt
        )
        .unwrap())
    }

    #[test]
    pub fn generates_correct_response_insufficient_exemption() {
        let hlt: HazardLookupTable = serde_json::from_str(COMPLICATED_HLT).unwrap();
        let metadata = Metadata {
            hlt_index: 198,
            an_subindex: 0,
            an_likelihood: half::f16::from_f32(0.5),
            provenance: Provenance::AAWildType,
            reverse_screened: false,
            is_common: false,
        };

        let exemption = Organism {
            name: "Othertoxin".into(),
            sequences: vec![SequenceIdentifier::Id(
                GenbankId::try_from("NC_00003".to_owned()).unwrap(),
            )],
        };

        insta::assert_yaml_snapshot!(HdbResponse::with_hlt(
            metadata,
            Region::Us,
            &make_exemptions(vec![exemption]),
            &hlt
        )
        .unwrap())
    }

    #[test]
    fn removes_low_risk_tag() {
        let hlt: HazardLookupTable = serde_json::from_str(
            r#"{ "entries": {
            "0": { "id_groups": [
                [{"OrganismName": "Test"}, {"OrganismType": "Virus"}, {"Tag": "HumanToHuman"}],
                [{"OrganismName": "Test"}, {"OrganismType": "Virus"}, {"Tag": "SdnaLowRiskDNA"}]
            ] }
        } }"#,
        )
        .unwrap();

        fn run_tags(
            hlt_index: u32,
            an_subindex: u8,
            hlt: &HazardLookupTable,
        ) -> (Vec<Tag>, Vec<Tag>) {
            let metadata = Metadata {
                hlt_index,
                an_subindex,
                an_likelihood: half::f16::from_f32(0.),
                provenance: Provenance::DnaNormal,
                reverse_screened: false,
                is_common: false,
            };

            let resp =
                HdbResponse::with_hlt(metadata, Region::All, &Default::default(), hlt).unwrap();
            let mut possible_tags: Vec<Tag> = resp
                .organisms
                .into_iter()
                .flat_map(|o| o.tags.into_iter())
                .collect();
            possible_tags.sort();
            possible_tags.dedup();
            (resp.most_likely_organism.tags, possible_tags)
        }

        assert_eq!(
            run_tags(0, 0, &hlt),
            (vec![Tag::HumanToHuman], vec![Tag::HumanToHuman],)
        );

        assert_eq!(
            run_tags(0, 1, &hlt),
            (
                vec![Tag::RegulatedButPass],
                vec![Tag::HumanToHuman, Tag::RegulatedButPass],
            )
        );
    }

    #[test]
    fn common_tag() {
        fn run_with_common(is_common: bool) -> (Vec<Tag>, Vec<Tag>) {
            let hlt: HazardLookupTable = serde_json::from_str(
                r#"{ "entries": {
                "0": { "id_groups": [
                    [{"OrganismName": "Test"}, {"OrganismType": "Virus"}]
                ] }
            } }"#,
            )
            .unwrap();

            let metadata = Metadata {
                hlt_index: 0,
                an_subindex: 0,
                an_likelihood: half::f16::from_f32(0.),
                provenance: Provenance::DnaNormal,
                reverse_screened: false,
                is_common, // the value we're checking is propagated into tags
            };

            let resp =
                HdbResponse::with_hlt(metadata, Region::All, &Default::default(), &hlt).unwrap();
            let mut possible_tags: Vec<Tag> = resp
                .organisms
                .into_iter()
                .flat_map(|o| o.tags.into_iter())
                .collect();
            possible_tags.sort();
            possible_tags.dedup();

            (resp.most_likely_organism.tags, possible_tags)
        }

        assert_eq!(
            run_with_common(true),
            (vec![Tag::Common], vec![Tag::Common])
        );
        assert_eq!(run_with_common(false), (vec![], vec![]));
    }
}
