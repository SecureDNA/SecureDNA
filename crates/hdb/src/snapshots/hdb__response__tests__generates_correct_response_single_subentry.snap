---
source: crates/hdb/src/response.rs
expression: "HdbResponse::with_hlt(metadata, Some(Region::Us), &Default::default(),\n        &hlt).unwrap()"
---
synthesis_permission: denied
most_likely_organism:
  name: Nastytoxin
  organism_type: Toxin
  ans:
    - NC_00002
  tags:
    - SelectAgentHhs
organisms:
  - name: Nastytoxin
    organism_type: Toxin
    ans:
      - NC_00002
    tags:
      - SelectAgentHhs
an_likelihood: 0.5
provenance: AAWildType
reverse_screened: false
window_gap: 3
exempt: false

